chainer/testing/function.py

```diff
@@ -0,0 +1,530 @@
+
+        Returns: A tuple of grad_grad_inputs arrays.
+        """
+        grad_grad_inputs = tuple(
+            numpy.random.uniform(-1, 1, _.shape).astype(_.dtype)
+            for _ in inputs_template)
+        return grad_grad_inputs
+
+    @_copy_to_test_case()
+    def _to_noncontiguous_as_needed(self, contig_arrays):
+        if self.contiguous is None:
+            # non-contiguous
+            return array_module._as_noncontiguous_array(contig_arrays)
+        if self.contiguous == 'C':
+            # C-contiguous
+            return contig_arrays
+        assert False, (
+            'Invalid value of `contiguous`: {}'.format(self.contiguous))
+
+    @_copy_to_test_case()
+    def _generate_inputs(self):
+        inputs = self.generate_inputs()
+        _check_array_types(inputs, _cpu.CpuDevice(), 'generate_inputs')
+        return inputs
+
+    @_copy_to_test_case()
+    def _generate_grad_outputs(self, outputs_template):
+        grad_outputs = self.generate_grad_outputs(outputs_template)
+        _check_array_types(
+            grad_outputs, _cpu.CpuDevice(), 'generate_grad_outputs')
+        return grad_outputs
+
+    @_copy_to_test_case()
+    def _generate_grad_grad_inputs(self, inputs_template):
+        grad_grad_inputs = self.generate_grad_grad_inputs(inputs_template)
+        _check_array_types(
+            grad_grad_inputs, _cpu.CpuDevice(), 'generate_grad_grad_inputs')
+        return grad_grad_inputs
+
+    @_copy_to_test_case()
+    def _forward_expected(self, inputs):
+        outputs = self.forward_expected(inputs)
+        _check_array_types(outputs, _cpu.CpuDevice(), 'forward_expected')
+        return outputs
+
+    @_copy_to_test_case()
+    def _forward(self, inputs, backend_config):
+        assert all(isinstance(a, chainer.Variable) for a in inputs)
+        with backend_config:
+            outputs = self.forward(inputs, backend_config.device)
+        _check_variable_types(outputs, backend_config.device, 'forward')
+        if not isinstance(outputs, tuple):
+            outputs = outputs,
+        return outputs
+
+    @_copy_to_test_case(enable_attr_name='forward_test')
+    def test_forward(self, backend_config):
+        """Tests forward computation."""
+
+        self.backend_config = backend_config
+        self.before_test('test_forward')
+
+        cpu_inputs = self._generate_inputs()
+        inputs_copied = [_.copy() for _ in cpu_inputs]
+
+        # Compute expected outputs
+        cpu_expected = self._forward_expected(cpu_inputs)
+        inputs = backend_config.get_array(cpu_inputs)
+        inputs = self._to_noncontiguous_as_needed(inputs)
+
+        # Compute actual outputs
+        outputs = self._forward(
+            tuple(chainer.Variable(a) for a in inputs),
+            backend_config)
+
+        # Check inputs has not changed
+        for i in range(len(inputs)):
+            with FunctionTestError.raise_if_fail(
+                    'Input array has been modified during forward.'):
```

Comment:
How about including some additional information such as the index of the inputs or some of its metadata (array props) to make debugging easier.